.github/workflows/main.yml

```diff
@@ -35,18 +36,68 @@ jobs:
 
       - name: Save docker image
         run: |
-          docker save research-template | gzip > /tmp/research-template.tar.gz
+          docker save research-template | pigz --fast > /tmp/research-template.tar.gz
 
       - name: Upload docker image
         uses: actions/upload-artifact@v4
         with:
             name: research-template-image
             path: /tmp/research-template.tar.gz
+            # Disable compression; the file is already compressed
+            compression-level: 0
 
-  publish:
+  test-packages:
     needs: [build-and-test]
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Download Docker image
+      uses: actions/download-artifact@v4
+      with:
+        name: research-template-image
+        path: /tmp/image
+
+    - name: Import Docker image
+      run: docker load --input /tmp/image/research-template.tar.gz
+
+    - name: Tag Docker image for testing
+      run: docker tag "$IMAGE_NAME" "$PUBLIC_IMAGE_NAME:dev"
+
+    - name: Checkout research template repository
+      uses: actions/checkout@v4
+      with:
+        repository: opensafely/research-template
 
-    runs-on: ubuntu-22.04
+    - name: Checkout research-template-docker repository in subdirectory
+      uses: actions/checkout@v4
+      with:
+        path: 'research-template/research-template-docker'
+
+    - name: Install demjson package for jsonlint
+      # Necessary because jq doesn't yet have an option to strip comments.
+      # See https://github.com/jqlang/jq/issues/1571
+      run: |
+        sudo apt-get update
+        sudo apt-get install python3-demjson
+
+    - name: Amend research-template devcontainer.json to use dev image
+      run: |
+        jsonlint --allow comments --format-compactly ".devcontainer/devcontainer.json" |
+          jq --arg variable "$PUBLIC_IMAGE_NAME:dev" '.image = $variable' > "./devcontainer/devcontainer.json."
+        cat "./research-template/.devcontainer/devcontainer.json"
+
+    # It is necessary to run these tests in the dev container,
+    # because they currently require Docker to correctly discover
+    # the state of the Python packages in the Python action image.
+    - name: Build and run dev container test task
+      uses: devcontainers/ci@a56d055efecd725e8cfe370543b6071b79989cc8 # v0.3.1900000349
+      with:
+        runCmd: ./research-template/research-template-docker/tests/packages.sh
```

Comment:
thought: This is slightly annoying because it depends on the dev container setup.

I'm not sure if there's a clean, non-hacky way of comparing the Python packages installed otherwise. The issue is that the output of `pip freeze` in the virtual environment in the Docker image being tested differs from the `requirements.txt` in the Python action image repository. So we end up running `pip freeze` in both Docker images: the OpenSAFELY Python action image and this `research-template-docker` image.

* the name casing is different, although you could normalise this to lower case, and the 

We can possibly go for:

* either a less clean approach with having to put discrepancies in a list
* installing Docker inside Docker (the dev container feature script that does this is a big one though)
* or add a minimal specific dev container setup in this repository to remove that dependency

Reply:
Actually, most of these options are the default and can be removed:

```
{
    "name": "Test Docker package install",
    "image": "ghcr.io/opensafely-core/research-template:dev",
    "features": {
        "ghcr.io/devcontainers/features/docker-in-docker:2": {
            "dockerDashComposeVersion": "v2"
        }
    }
}
```

The only one that isn't is the `dockerDashComposeVersion`.

Reply:
Or am I half-asleep and we can just run this in Docker directly?

Reply:
> Or am I half-asleep and we can just run this in Docker directly?

We can, so did that instead.